### DISCUSSION TITLE
Newbie: How to handle custom date/time-based versioning for docker image?

### DISCUSSION THREAD

**Discussion by pmorch** [renovatebot/renovate#13020] [Request Help]
Hi,

I'm new to renovate. We have number of images, that get tags like this every time jenkins builds them:

    $branch-$dateTime-$shortSha

E.g.

    develop-2021-12-08_11_45_45-b5dc664
    develop-2021-12-07_17_53_12-10f5f62

We want renovate to update our ArgoCD GitOps manifests with the latest build every time anyone pushes to the `develop` branch (in this case).

I'm happy to assume there won't be two builds within the same second (I wish), so just taking the alphabetically largest one that starts with `develop-` would be fine with me.

I've looked at [Renovate Docs: Versioning](https://docs.renovatebot.com/modules/versioning/) and haven't really found any good way to "take the alphabetically largest one that starts with `develop-`".

[Regular Expression Versioning](https://docs.renovatebot.com/modules/versioning/#regular-expression-versioning) seems the closest fit, but even it says:

> `major`, `minor`, and `patch`: at least one of these must be provided. When determining whether a package has updates, these values will be compared in the standard semantic versioning fashion. If any of these fields are omitted, they will be treated as if they were 0 -- in this way, you can describe versioning schemes with up to three incrementing values.

So it still assumes something semver-y. I guess `develop` would be the `compatibility` field, but then what?

[semver 2.0 ](https://semver.org/) (== "standard semantic versioning"?) states that X, Y and Z in X.Y.Z 

> MUST NOT contain leading zeroes

which dates and times clearly do.

I also looked at "Loose Versioning", but it doesn't have a concept of `compatibility` and so I rejected it, because I only want to consider images for the same `$branch`.

Surely I can't be the first person to come across date/timestamp based versions...

**Comment by rarkins** [comment]
There might be more than one way to do this.. hopefully not zero.

Perhaps loose versioning will work but you also need to use allowedVersions with a regex to enforce the develop- prefix.

Alternatively extractVersion with loose versioning might also work. Either way you will need to specify that developer- branch in your config - it won't be "automatically" detected.

**Reply by pmorch** [reply]
Thanks! extractVersion seems to be the part of the puzzle I was missing. Thanks!

**Reply by pmorch** [reply]
## Short Version:

On a second inspection, I'm not sure `extractVersion` or `loose` will work. I hope I'm wrong.

I think I have it working with a `versioningTemplate` using `regex` and just the `patch` field, but it was not intuitive to get working and I still don't know _why_ it works. But it seems to... E.g. it didn't work if i used `major` instead of `patch` but really I just wanted to use "the alphabetically largest one that starts with `develop-`" - but I didn't succeed with that.

On one hand, it seems to be working for me now, and so we could happily move on. My itch is scratched. On the other hand, I would've preferred a better user experience (e.g. understand why it works) and I'm happy to engage in a discussion about it. But if you consider this a waste of your time, we'll just leave it as-is (until I discover a corner case that still doesn't work).

## Longer Version

We have dependencies in a `dynimages.libjsonnet` [jsonnet](https://jsonnet.org/) file like this:
```
{
    // renovate: my-timestamp-image
    'harbor.example.dk/dev/my-keycloak': 'develop-2020-11-05T09-16-26Z-deadbeef',
    // renovate: my-timestamp-image
    'harbor.example.dk/dev/my-app': 'develop-2021-11-05T09-16-26Z-abc123',
}
```

, tags like these in our own docker registry (similar to the format [minio uses](https://hub.docker.com/r/minio/minio/tags)):
```
develop-2021-12-08T11-45-45-b5dc664
develop-2021-12-07T17-53-12-10f5f62
```

And this incantation seems to work for me:
```
  "regexManagers": [
    {
      "fileMatch": ["^lib/dynimages.libjsonnnet$"],
      "matchStrings": [
        "(?:\\/\\/|#)\\s*renovate:\\s+my-timestamp-image\\n\\s*[\"'](?<depName>[^\"']+)[\"']\\s*:\\s*[\"'](?<currentValue>[^\"']+)[\"']"
      ],
      "versioningTemplate": "regex:^(?<compatibility>.*)-(?<patch>20\\d\\d)-.*$",
      "datasourceTemplate": "docker",
    },
  ],
```

Why does it work, when the `versioningTemplate` only matches the year as the `patch` value and the rest of the version field is matched but not stored in a capture group? I have no idea...

### About `extractVersion`

The [docs](https://docs.renovatebot.com/configuration-options/#extractversion) say:

> Only use this config option when the raw version strings from the datasource do not match the expected format that you need in your package file. 

But I want to use the same version strings in my "package" file that are used in the datasource. From reading the doc above, it doesn't sound like `extractVersion` is the right for this use case.

Also, when I try to use it with:

```
  "regexManagers": [
    {
      "fileMatch": ["^lib/dynimages.libjsonnnet$"],
      "matchStrings": [
        "(?:\\/\\/|#)\\s*renovate:\\s+my-timestamp-image\\n\\s*[\"'](?<depName>[^\"']+)[\"']\\s*:\\s*[\"'](?<currentValue>[^\"']+)[\"']"
      ],
      "versioningTemplate": "loose",
      "extractVersion": "^develop-(?<version>.*)$",
      "datasourceTemplate": "docker",
    },
  ],
```

I get:

```
       "error": {
         "validationSource": ".gitlab/renovate.json5",
         "validationError": "The renovate configuration file contains some invalid settings",
         "validationMessage": "Regex Manager contains disallowed fields: extractVersion",
         "message": "config-validation",
         <snip>
```

So I can't use `extractVersion` with `regexManagers`, but as far as I can see, I need a `regexManager` because of my proprietary  `dynimages.libjsonnet` file.

## About `loose`

As I read `loose`'s `parse` function:

https://github.com/renovatebot/renovate/blob/77aa6475f5e9bffb7034d261998f4960fdf403d0/lib/versioning/loose/index.ts#L10-L21

the version absolutely must start with `versionPattern`, so I'm outta luck with `develop-2021-12-08T11-45-45-b5dc664`. I _have to_ use `regex`.

**Reply by cascandaliato** [reply]
> Why does it work, when the versioningTemplate only matches the year as the patch value and the rest of the version field is matched but not stored in a capture group? I have no idea...

@pmorch, is Renovate triggering a version update even when the patch/year number is the same between two versions?

**Comment by akhan23wgu** [comment]
I have a similar issue. I almost gave up after a day of trying to get regex to detect `1.31.2.6810-a607d384f` as a valid version.

NOTHING worked. I regex101.com’d and attempted to use EVERY working solution. Instant fail once I made my commit for each one. There’s definitely something up with how the `major` group works.

Here’s my working solution:
`      "versioningTemplate": "regex:^(?<minor>\\d+)\\.(?<build>\\d+)\\.(?<patching>\\d+)\\.(?<revision>\\d+)-.*$",`

Variations that passed regex but failed in renovate:

```
^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\.(?<build>\\d+(.*[a-z0-9].*?)$
^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)\\.(?<build>\\d+)([-a-zA-Z0-9]{10})$
```
